domain/src/main/java/org/oppia/android/domain/question/QuestionAssessmentCalculation.kt

```diff
@@ -0,0 +1,105 @@
+package org.oppia.android.domain.question
+
+import org.oppia.android.app.model.FractionGrade
+import org.oppia.android.app.model.UserAssessmentPerformance
+import javax.inject.Inject
+import javax.inject.Singleton
+
+/**
+ * Private mutable class that stores a grade as a fraction.
+ */
+private class MutableFractionGrade(var numerator: Int, var denominator: Int)
+
+/**
+ * Private class that computes the state of the user's performance at the end of a practice session.
+ * This class can exist across multiple training session instances, but calling code is responsible
+ * for ensuring it is properly reset.
+ */
+@Singleton
+internal
+class QuestionAssessmentCalculation @Inject constructor(
+  @ViewHintScorePenalty private val viewHintPenalty: Int,
+  @WrongAnswerScorePenalty private val wrongAnswerPenalty: Int,
+  @MaxScorePerQuestion private val maxScorePerQuestion: Int
+) {
+  private lateinit var skillIdList: List<String>
+  private lateinit var questionSessionMetrics: List<QuestionSessionMetrics>
+  private lateinit var totalScore: MutableFractionGrade
+  private lateinit var scorePerSkillMapping: MutableMap<String, MutableFractionGrade>
+
+  /** Initialize member fields. */
+  internal fun initialize(
+    skillIdList: List<String>,
+    questionSessionMetrics: List<QuestionSessionMetrics>
+  ) {
+    this.skillIdList = skillIdList
+    this.questionSessionMetrics = questionSessionMetrics
+    this.totalScore =
+      MutableFractionGrade(0, 0)
+    createScorePerSkillMapping()
+  }
+
+  /** Initialize the scorePerSkillMapping member field. */
+  private fun createScorePerSkillMapping() {
+    scorePerSkillMapping = mutableMapOf()
+    for (skillId in skillIdList) {
+      scorePerSkillMapping[skillId] = MutableFractionGrade(0, 0)
+    }
+  }
+
+  /** Calculate the user's overall score and score per skill for this practice session. */
+  private fun calculateScores() {
+    for (questionMetric in questionSessionMetrics) {
+      val totalHintsPenalty =
+        questionMetric.numberOfHintsUsed * viewHintPenalty
+      val totalWrongAnswerPenalty =
+        (questionMetric.numberOfAnswersSubmitted - 1) * wrongAnswerPenalty
+      var questionScore = 0
+      if (!questionMetric.didViewSolution) {
+        questionScore =
+          maxOf(
+            maxScorePerQuestion - totalHintsPenalty - totalWrongAnswerPenalty,
+            questionScore
+          )
+      }
+      this.totalScore.numerator += questionScore
+      this.totalScore.denominator += maxScorePerQuestion
+      for (linkedSkillId in questionMetric.question.linkedSkillIdsList) {
+        if (!scorePerSkillMapping.containsKey(linkedSkillId)) continue
+        scorePerSkillMapping[linkedSkillId]!!.numerator += questionScore
+        scorePerSkillMapping[linkedSkillId]!!.denominator += maxScorePerQuestion
```

Comment:
This can be made safer by using a scoping function & avoiding the '!!':

```suggestion
        scorePerSkillMapping[linkedSkillId]?.let { mutableGrade ->
          mutableGrade.numerator += questionScore
          mutableGrade.denominator += maxScorePerQuestion
        }
```

Note that the current solution technically has a race condition since controllers can be called by multiple threads & it's assuming nothing changes between the containsKey and the retrieval (which is actually probably true if you initialize the map in the class's constructor per my factory comment above, but it's still good to address this part as well).

Reply:
Fixed